docs/design-documents/20220408-loading-pipeline-configuration-files.md

```diff
@@ -0,0 +1,282 @@
+- We shouldn't implement key management or encryption for secrets.
+- We shouldn't care or design for how environment variables are set or managed.
+
+## Scope
+
+We are strictly concerned with the loading of pipeline files at start time and 
+how those files are handled through Conduit's full operation cycle.
+
+Configuration file schema and other file-level design choices are out of scope.
+
+This document doesn't make any definitions or statements about how the UI should
+handle these changes. Those should be reserved for later design documents and
+discussion.
+
+## Key Terms 
+
+**pipeline** - a Conduit pipeline with `n` number of Source connectors, 
+`m` number of Processors, and `o` number of Destination connectors.
+
+**production mode** - a mode of operation where Conduit pipelines must be
+statically defined and cannot change after Conduit starts. No UI is served in
+production mode, and only pipeline lifecycle methods can be called over the
+API.
+
+## Context 
+
+Running Conduit in production environments requires a strict set of
+guarantees. Manually configuring a Conduit server in a GUI and keeping those
+high standards places more work on the end user, reduces confidence in the
+pipeline's integrity, and in the worst case could result in corrupted or lost
+data.
+
+This document expands on previous discussions to create a production mode of
+Conduit that either doesn't serve the application GUI or serves a functionally 
+limited version of the GUI instead.
+
+Conduit Issues #32 and #33 describe two different features to address some of 
+these concerns, but there are additional design decisions to make.
+
+## Design 
+
+We propose the creation of a pipeline configuration file that contains all of
+the configuration values needed for `n` number of full pipelines.
+
+We propose naming this file `pipelines.yml` and it's default location should be
+at the root of the project. The format of the file is to be determined, though
+`yaml` has been proposed previously.
+
+If this file is passed at start time, we propose that Conduit run in production 
+mode. Production mode initially will not serve the UI. In the future, it would 
+serve a read-only optimized view of Conduit.
+
+Configuration must securely accept secret information. We don't want to solve
+the problem of secrets management, but we must be in the business of secret
+concealment. Additionally, we already allow environment variables so we must 
+continue handling those correctly. 
+
+We propose merging CLI flags, environment variables, and the user's provided 
+configuration file, in that order, at runtime into template variables in the 
+configuration file with their secret values at runtime.
+
+We propose merging environment variables and CLI flags into a set of values 
+that is then injected into the provided set of pipeline files.
+
+We propose supporting Go templating in all pipeline files to allow for dynamic
+injection of those environment variables at runtime.
```

Comment:
Can you explain what you had in mind with Go templating?

Reply:
I imagine something like [benjohnson/tmpl](https://github.com/benbjohnson/tmpl) but custom. Go style template variables would be used to declare where a value should be used, and then at start-time the corresponding variable would be injected.